CRM/Extension/Browser.php

```diff
@@ -157,21 +157,31 @@ public function getExtensions() {
 
   /**
    * Get a description of a particular extension.
+   * This function is an optimised form of getExtensions() - it returns the
+   * extension's details as soon as it's found without caching anything.
    *
    * @param string $key
    *   Fully-qualified extension name.
    *
    * @return CRM_Extension_Info|NULL
    */
   public function getExtension($key) {
-    // TODO optimize performance -- we don't need to fetch/cache the entire repo
-    $exts = $this->getExtensions();
-    if (array_key_exists($key, $exts)) {
-      return $exts[$key];
+    if (!$this->isEnabled() || count($this->checkRequirements())) {
+      return NULL;
     }
-    else {
+
+    $remote = $this->_discoverRemote();
```

Comment:
As a preface: my gut says that the "TODO" comment is old/stale -- at the time it was written, `_discoverRemote()` was very expensive (crawling a remote webpage and downloading separate files for each extension). That part was previously optimized by FF crew at an Edale sprint. (It now fetches one JSON document).

> When I measured it it performed faster, and will definitely use less memory.

Just a fair warning -- the cache timeout in `_discoverRemote()` is 180 seconds. This might make it hard to reliably compare the performance of different codes/algorithms. (Without carefully controlling the network IO from `_discoverRemote()`, you'd suffer a big+irrelevant performance hit every couple minutes.) For testing internal performance issues, I'd bump up the TTL real high (like 86400 seconds).

Never-the-less, the memory usage is an intriguing point. The full JSON doc for all developmental extensions is ~120kb. Not a huge deal by itself... but `$this->_remotesDiscovered` looks a little mismanaged to me. (It stores the data as a property, but never uses it a second time.) If there's a use-case which actually does call `_discoverRemote()` or `getExtensions()` several times, then you could imagine it wasting time on memory-management. (But if that is an issue... then I think the fix would be to improve reuse of `$this->_remotesDiscovered`.)